src/Passes/Source/Rules/Factory.cpp

```diff
@@ -0,0 +1,334 @@
+
+                    llvm::FunctionType* fnc_type = llvm::FunctionType::get(return_type, types, false);
+                    function                     = llvm::Function::Create(
+                        fnc_type, llvm::Function::ExternalLinkage, "__quantum__qis__mz__body", module);
+                }
+
+                // Ensuring we are inserting after the instruction being deleted
+                builder.SetInsertPoint(llvm::dyn_cast<llvm::Instruction>(val)->getNextNode());
+
+                builder.CreateCall(function, arguments);
+
+                // Replacing the instruction with new instruction
+                // TODO(tfr): (tfr): insert instruction before and then replace, with new call
+                replacements.push_back({llvm::dyn_cast<Instruction>(val), instr});
+
+                return true;
+            };
+
+        addRule({call("__quantum__qis__m__body", "qubit"_cap = _), std::move(replace_measurement)});
+    }
+
+    void RuleFactory::optimiseBranchQuatumOne()
+    {
+        auto replace_branch_positive = [](Builder& builder, Value* val, Captures& cap, Replacements& replacements) {
+            auto result = cap["result"];
+            auto cond   = llvm::dyn_cast<llvm::Instruction>(cap["cond"]);
+            // Replacing result
+            auto                      module   = llvm::dyn_cast<llvm::Instruction>(val)->getModule();
+            auto                      function = module->getFunction("__quantum__qir__read_result");
+            std::vector<llvm::Value*> arguments;
+            arguments.push_back(result);
+
+            if (!function)
+            {
+                std::vector<llvm::Type*> types;
+                types.resize(arguments.size());
+                for (uint64_t i = 0; i < types.size(); ++i)
+                {
+                    types[i] = arguments[i]->getType();
+                }
+
+                auto return_type = llvm::Type::getInt1Ty(val->getContext());
+
+                llvm::FunctionType* fnc_type = llvm::FunctionType::get(return_type, types, false);
+                function                     = llvm::Function::Create(
+                    fnc_type, llvm::Function::ExternalLinkage, "__quantum__qir__read_result", module);
+            }
+
+            builder.SetInsertPoint(llvm::dyn_cast<llvm::Instruction>(val));
+            auto new_call = builder.CreateCall(function, arguments);
+            new_call->takeName(cond);
+
+            for (auto& use : cond->uses())
+            {
+                llvm::User* user = use.getUser();
+                user->setOperand(use.getOperandNo(), new_call);
+            }
+            cond->replaceAllUsesWith(new_call);
+
+            // Deleting the previous condition and function to fetch one
+            replacements.push_back({cond, nullptr});
+            replacements.push_back({cap["one"], nullptr});
+
+            return false;
+        };
+
+        /*
```

Comment:
That should be removed.

Reply:
I actually don't think it should be removed. Instead it should be expanded as it is quite helpful to have an example IR that it is supposed to match next to the pattern definition.

Reply:
Then it needs to be morphed into a proper comment. As is, it is not clear what to take away from this comment.